src/main/java/com/atlassian/db/replica/internal/DecisionAwareReference.java

```diff
@@ -0,0 +1,27 @@
+package com.atlassian.db.replica.internal;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class DecisionAwareReference<T> extends LazyReference<T> {
+    private final AtomicReference<RouteDecisionBuilder> decisionBuilder = new AtomicReference<>();
+
+    public T get(RouteDecisionBuilder decisionBuilder) {
+        if (!isInitialized()) {
+            this.decisionBuilder.compareAndSet(null, decisionBuilder);
+        }
+        return super.get();
```

Comment:
So when `isInitialized()`, then we ignore the `decisionBuilder`. Intentional?

Reply:
We need to store the cause when the connection is initialised. Otherwise, it will not be a cause. OTOH we can ignore it as CAS will change the value only once.

Reply:
Ok, simplified a little: https://github.com/wyrzyk/db-replica/pull/4